.NET Framework exception

OrionNebula opened the pull request:
CultureInfo.GetCultureInfo(string) sometimes throws an exception when a culture isn't found, so I need to catch that.
Fixed some locale names that CultureInfo didn't like

OrionNebula commented on examples/ceilidh-dummy/index.ts:
```diff
@@ -1,20 +1,22 @@
+import { InnerProviderContract } from 'io.ceilidh.dummy'
 
-export class InnerProvider extends InnerConsumerContract {
+export class InnerProvider extends InnerProviderContract {
```
No, provider classes implement abstract provider contracts

JustANull replied:
Ah, I was confused by the name.

OrionNebula commented on examples/ceilidh-dummy/io.ceilidh.dummy.d.ts:
```diff
@@ -1,5 +1,5 @@
 declare module 'io.ceilidh.dummy' {
```
This lets us import provider contracts in a type-safe way after we patch Module._load

OrionNebula commented on examples/ceilidh-remote-intent/package.json:
```diff
@@ -0,0 +1,10 @@
+  "version": "0.0.0",
+  "description": "A dummy plugin demonstrating how to create plugins for Ceilidh.",
+  "main": "dist/index.js",
+  "manifest": "dist/manifest.js",
```
package.json has meaning to node itself (specifically main) that I rely on, so package.json must be present regardless

OrionNebula commented on package.json:
```diff
@@ -9,7 +9,7 @@
     "lint:fix": "tslint --fix --config tslint.json --project src",
     "prebuild": "npm run lint:fix && rimraf core",
-    "build": "tsc --project src && tsc --project examples/ceilidh-dummy",
+    "build": "tsc --project src && tsc --project examples/ceilidh-dummy && tsc --project examples/ceilidh-remote-intent",
```
It'd have to be cross-platform

OrionNebula commented on src/plugins/PluginLoader.ts:
```diff
@@ -136,7 +145,7 @@ interface ProvidersConsumers<T> {
  * @param pluginPath The path to the plugin to load
  */
-export function loadPluginFinalPass (mod: ExpandedModule): ProvidersConsumers<{}> {
+export function loadPluginFinalPass (mod: ExpandedModule): FinalPass<{}> {
```
It's just an empty object

OrionNebula commented on src/plugins/WorkerBootstrap.ts:
```diff
@@ -22,10 +27,40 @@ export function loadIntent (pluginPaths: string[], extraProviderClasses?: Provid
+        const consumer: { [key: string]: any } = new consumerClass([
+          ...(extraProviderClasses || []),
+          ...y.providers
+        ].filter(prov => prov.prototype instanceof consumerClass.providerContract)
```
No, the shape is [] because of the spread operator (...)

JustANull replied:
Ah, ok.

OrionNebula commented on src/rpc/Routing.ts:
```diff
@@ -0,0 +1,445 @@
+      }
+
+      const child = intentProcs[intent]
+      const id = responseId = (responseId + 1) % 1024 // Arbitrary number, I just don't want it to loop
```
ID lets us actually match a response message to a request

OrionNebula commented:
@JustANull 
I made the requested changes, sort of the global dispatch system for handling invoke responses.

OrionNebula commented on src/core.ts:
```diff
@@ -28,25 +29,37 @@ interface IntentGroups {
-    for (const key in groups) { // for each intent group, spawn a new fork
-      if (!groups.hasOwnProperty(key)) break
+    return Promise.all(Object.keys(groups).map(key => {
+      if (!groups.hasOwnProperty(key)) return
```
Well it's not just over the properties, because I also need the key associated with that property. I'm pretty sure this is the best way to do it? But not certain.

OrionNebula commented on src/rpc/Routing.ts:
```diff
@@ -0,0 +1,445 @@
+    return new Promise((resolve, reject) => {
+      const id = responseId = (responseId + 1) % 1024 // Generate a reasonably unique response ID
+
+      core.once('message', handleMessage) // Setup the 'once' loop
```
Resolved by using `onceWhen`

OrionNebula commented on src/rpc/Routing.ts:
```diff
@@ -0,0 +1,445 @@
+   */
+  export function invokeRemoteIpcPort (intent: string, action: string, core: CoreIpcChannel): Promise<Duplex> {
+    return new Promise((resolve, reject) => {
+      const id = responseId = (responseId + 1) % 1024 // Generate a reasonably unique response ID
```
First one is resolved with `onceWhen`
Second one uses that new thing, and invokes are expected to resolve quickly enough that we won't hit the loop point before we hit it. (especially because port invokes don't reserve a responseId)

OrionNebula committed:
.NET Framework exception

CultureInfo.GetCultureInfo(string) sometimes throws an exception when a culture isn't found, so I need to catch that.
Fixed some locale names that CultureInfo didn't like
